File: legacy/src/com/cattailsw/nanidroid/install/NarStagedTreeInventory.java
package com.cattailsw.nanidroid.install;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Pure immutable adapter from native staged-tree facts to the reviewed policy.
 *
 * <p>Fingerprint v1 authenticates target, storage-root identity, logical paths
 * and copied-byte digests, but intentionally excludes blob ordinals. The
 * immutable manifest and ordinal inventory must remain inseparable in the
 * later capability. D9b3 must treat process-death orphans as discard-only.
 */
final class NarStagedTreeInventory {
    private static final int MAX_ENTRIES = 10000;
    private static final int DIGEST_BYTES = 32;

    private NarStagedTreeInventory() {}

    enum Error { NATIVE, POLICY }

    static final class Description {
        private final long storageDevice;
        private final long storageInode;
        private final String[] paths;
        private final int[] types;
        private final long[] sizes;
        private final int[] ordinals;
        private final byte[] digests;

        Description(long storageDevice, long storageInode,
                String[] paths, int[] types, long[] sizes,
                int[] ordinals, byte[] digests) {
            this.storageDevice = storageDevice;
            this.storageInode = storageInode;
            this.paths = paths == null ? null : paths.clone();
            this.types = types == null ? null : types.clone();
            this.sizes = sizes == null ? null : sizes.clone();
            this.ordinals = ordinals == null ? null : ordinals.clone();
            this.digests = digests == null ? null : digests.clone();
        }
    }

    static final class Entry {
        private final String path;
        private final NarGhostTreePolicy.Type type;
        private final long size;
        private final int blobOrdinal;
        private final byte[] sha256;

        private Entry(String path, NarGhostTreePolicy.Type type,
                long size, int blobOrdinal, byte[] sha256) {
            this.path = path;
            this.type = type;
            this.size = size;
            this.blobOrdinal = blobOrdinal;
            this.sha256 = sha256 == null ? null : sha256.clone();
        }

        String path() { return path; }
        NarGhostTreePolicy.Type type() { return type; }
        long size() { return size; }
        int blobOrdinal() { return blobOrdinal; }
        byte[] sha256() {
            return sha256 == null ? null : sha256.clone();
        }
    }

    static final class Result {
        private final NarGhostTreePolicy.Manifest manifest;
        private final List<Entry> entries;
        private final Error error;
        private final String detail;

        private Result(NarGhostTreePolicy.Manifest manifest,
                List<Entry> entries, Error error, String detail) {
            this.manifest = manifest;
            this.entries = entries;
            this.error = error;
            this.detail = detail;
        }

        private static Result success(
                NarGhostTreePolicy.Manifest manifest,
                List<Entry> entries) {
            return new Result(manifest, Collections.unmodifiableList(
                    new ArrayList<Entry>(entries)), null, "");
        }

        private static Result failure(Error error, String detail) {
            return new Result(null, Collections.<Entry>emptyList(),
                    error, detail);
        }

        boolean isSuccess() { return manifest != null; }
        NarGhostTreePolicy.Manifest manifest() { return manifest; }
        List<Entry> entries() { return entries; }
        Error error() { return error; }
        String detail() { return detail; }
    }

    static Result absent(
            String target, long storageDevice, long storageInode) {
        return build(target, storageDevice, storageInode,
                NarGhostTreePolicy.State.ABSENT, null);
    }

    static Result present(String target, Description description) {
        if (description == null) {
            return Result.failure(Error.NATIVE, "description");
        }
        return build(target, description.storageDevice,
                description.storageInode,
                NarGhostTreePolicy.State.PRESENT, description);
    }

    private static Result build(String target,
            long storageDevice, long storageInode,
            NarGhostTreePolicy.State state,
            Description description) {
        try {
            Prepared prepared = state == NarGhostTreePolicy.State.ABSENT
                    ? Prepared.absent() : prepare(description);
            if (prepared == null) {
                return Result.failure(Error.NATIVE, "inventory");
            }
            NarGhostTreePolicy.Result policy = NarGhostTreePolicy.build(
                    target, identity(storageDevice, storageInode),
                    state, prepared.policyEntries);
            if (!policy.isSuccess()) {
                return Result.failure(
                        Error.POLICY, policy.getError().name());
            }
            ArrayList<Entry> entries = new ArrayList<Entry>();
            for (NarGhostTreePolicy.Entry entry
                    : policy.getManifest().getEntries()) {
                Entry item = prepared.byPath.get(entry.getPath());
                if (item == null) {
                    return Result.failure(
                            Error.NATIVE, "inventory mapping");
                }
                entries.add(item);
            }
            return Result.success(policy.getManifest(), entries);
        } catch (RuntimeException error) {
            return Result.failure(Error.NATIVE, "inventory");
        }
    }

    private static final class Prepared {
        private final List<NarGhostTreePolicy.InputEntry> policyEntries;
        private final Map<String, Entry> byPath;

        private Prepared(
                List<NarGhostTreePolicy.InputEntry> policyEntries,
                Map<String, Entry> byPath) {
            this.policyEntries = policyEntries;
            this.byPath = byPath;
        }

        private static Prepared absent() {
            return new Prepared(
                    Collections.<NarGhostTreePolicy.InputEntry>emptyList(),
                    Collections.<String, Entry>emptyMap());
        }
    }

    private static Prepared prepare(Description value) {
        if (value.paths == null || value.types == null
                || value.sizes == null || value.ordinals == null
                || value.digests == null) {
            return null;
        }
        int count = value.paths.length;
        if (count > MAX_ENTRIES || value.types.length != count
                || value.sizes.length != count
                || value.ordinals.length != count
                || value.digests.length != count * DIGEST_BYTES) {
            return null;
        }
        ArrayList<NarGhostTreePolicy.InputEntry> policy =
                new ArrayList<NarGhostTreePolicy.InputEntry>(count);
        HashMap<String, Entry> inventory =
                new HashMap<String, Entry>();
        boolean[] seen = new boolean[count];
        int files = 0;
        for (int index = 0; index < count; index++) {
            String path = value.paths[index];
            NarRelativePathPolicy.Result normalized =
                    NarRelativePathPolicy.normalize(path);
            if (!normalized.isSuccess()) return null;
            if (value.types[index] == 2) {
                if (value.sizes[index] != 0
                        || value.ordinals[index] != -1
                        || !zeroDigest(value.digests, index)) {
                    return null;
                }
                Entry entry = new Entry(normalized.getNormalized(),
                        NarGhostTreePolicy.Type.DIRECTORY,
                        0, -1, null);
                inventory.put(entry.path, entry);
                policy.add(
                        NarGhostTreePolicy.InputEntry.directory(path));
            } else if (value.types[index] == 1
                    && value.sizes[index] >= 0
                    && value.ordinals[index] >= 0
                    && value.ordinals[index] < count
                    && !seen[value.ordinals[index]]) {
                seen[value.ordinals[index]] = true;
                files++;
                byte[] digest = new byte[DIGEST_BYTES];
                System.arraycopy(value.digests,
                        index * DIGEST_BYTES, digest, 0, DIGEST_BYTES);
                Entry entry = new Entry(normalized.getNormalized(),
                        NarGhostTreePolicy.Type.FILE,
                        value.sizes[index],
                        value.ordinals[index], digest);
                inventory.put(entry.path, entry);
                policy.add(NarGhostTreePolicy.InputEntry.file(
                        path, value.sizes[index], digest));
            } else {
                return null;
            }
        }
        for (int ordinal = 0; ordinal < files; ordinal++) {
            if (!seen[ordinal]) return null;
        }
        for (int ordinal = files; ordinal < seen.length; ordinal++) {
            if (seen[ordinal]) return null;
        }
        return new Prepared(policy, inventory);
    }

    private static boolean zeroDigest(byte[] values, int index) {
        int start = index * DIGEST_BYTES;
        for (int offset = 0; offset < DIGEST_BYTES; offset++) {
            if (values[start + offset] != 0) return false;
        }
        return true;
    }

    private static byte[] identity(long device, long inode) {
        byte[] result = new byte[16];
        putLong(result, 0, device);
        putLong(result, 8, inode);
        return result;
    }

    private static void putLong(
            byte[] target, int offset, long value) {
        for (int index = 7; index >= 0; index--) {
            target[offset + index] = (byte) value;
            value >>>= 8;
        }
    }
}

File: src/com/cattailsw/nanidroid/install/NarStagedTreeInventory.kt
package com.cattailsw.nanidroid.install

import java.util.Collections

/**
 * Pure immutable adapter from native staged-tree facts to the reviewed policy.
 *
 * Fingerprint v1 authenticates target, storage-root identity, logical paths
 * and copied-byte digests, but intentionally excludes blob ordinals. The
 * immutable manifest and ordinal inventory must remain inseparable in the
 * later capability. D9b3 must treat process-death orphans as discard-only.
 */
internal object NarStagedTreeInventory {
    private const val MAX_ENTRIES = 10_000
    private const val DIGEST_BYTES = 32

    enum class Error { NATIVE, POLICY }

    class Description(
        internal val storageDevice: Long,
        internal val storageInode: Long,
        paths: Array<String>?,
        types: IntArray?,
        sizes: LongArray?,
        ordinals: IntArray?,
        digests: ByteArray?,
    ) {
        internal val paths = paths?.clone()
        internal val types = types?.clone()
        internal val sizes = sizes?.clone()
        internal val ordinals = ordinals?.clone()
        internal val digests = digests?.clone()
    }

    class Entry internal constructor(
        // This is deliberately nullable at the raw-facts boundary.  The
        // overlay validator—not this value object—maps malformed supplied
        // facts to MALFORMED_BASELINE, matching the Java implementation.
        private val pathValue: String?,
        private val typeValue: NarGhostTreePolicy.Type,
        private val sizeValue: Long,
        private val blobOrdinalValue: Int,
        sha256: ByteArray?,
    ) {
        private val sha256Value = sha256?.clone()

        fun path(): String? = pathValue

        fun type(): NarGhostTreePolicy.Type = typeValue

        fun size(): Long = sizeValue

        fun blobOrdinal(): Int = blobOrdinalValue

        fun sha256(): ByteArray? = sha256Value?.clone()
    }

    class Result private constructor(
        private val manifestValue: NarGhostTreePolicy.Manifest?,
        private val entriesValue: List<Entry>,
        private val errorValue: Error?,
        private val detailValue: String,
    ) {
        fun isSuccess(): Boolean = manifestValue != null

        fun manifest(): NarGhostTreePolicy.Manifest? = manifestValue

        fun entries(): List<Entry> = entriesValue

        fun error(): Error? = errorValue

        fun detail(): String = detailValue

        companion object {
            fun success(
                manifest: NarGhostTreePolicy.Manifest,
                entries: List<Entry>,
            ): Result = Result(
                manifest,
                Collections.unmodifiableList(ArrayList(entries)),
                null,
                "",
            )

            fun failure(error: Error, detail: String): Result =
                Result(null, Collections.emptyList(), error, detail)
        }
    }

    fun absent(target: String?, storageDevice: Long, storageInode: Long): Result =
        build(target, storageDevice, storageInode, NarGhostTreePolicy.State.ABSENT, null)

    fun present(target: String?, description: Description?): Result {
        if (description == null) return Result.failure(Error.NATIVE, "description")
        return build(
            target,
            description.storageDevice,
            description.storageInode,
            NarGhostTreePolicy.State.PRESENT,
            description,
        )
    }

    private fun build(
        target: String?,
        storageDevice: Long,
        storageInode: Long,
        state: NarGhostTreePolicy.State,
        description: Description?,
    ): Result = try {
        val prepared = if (state == NarGhostTreePolicy.State.ABSENT) {
            Prepared.absent()
        } else {
            prepare(description)
        }
        if (prepared == null) return Result.failure(Error.NATIVE, "inventory")
        val policy = NarGhostTreePolicy.build(
            target,
            identity(storageDevice, storageInode),
            state,
            prepared.policyEntries,
        )
        if (!policy.isSuccess()) {
            return Result.failure(Error.POLICY, policy.error!!.name)
        }
        val entries = ArrayList<Entry>()
        for (entry in policy.manifest!!.entries) {
            val item = prepared.byPath[entry.path]
                ?: return Result.failure(Error.NATIVE, "inventory mapping")
            entries.add(item)
        }
        Result.success(policy.manifest, entries)
    } catch (_: RuntimeException) {
        Result.failure(Error.NATIVE, "inventory")
    }

    private class Prepared(
        val policyEntries: List<NarGhostTreePolicy.InputEntry>,
        val byPath: Map<String, Entry>,
    ) {
        companion object {
            fun absent(): Prepared = Prepared(Collections.emptyList(), Collections.emptyMap())
        }
    }

    private fun prepare(value: Description?): Prepared? {
        if (value == null || value.paths == null || value.types == null ||
            value.sizes == null || value.ordinals == null || value.digests == null
        ) return null
        val paths = value.paths
        val types = value.types
        val sizes = value.sizes
        val ordinals = value.ordinals
        val digests = value.digests
        val count = paths.size
        if (count > MAX_ENTRIES || types.size != count || sizes.size != count ||
            ordinals.size != count || digests.size != count * DIGEST_BYTES
        ) return null

        val policy = ArrayList<NarGhostTreePolicy.InputEntry>(count)
        val inventory = HashMap<String, Entry>()
        val seen = BooleanArray(count)
        var files = 0
        for (index in 0 until count) {
            val path = paths[index]
            val normalized = NarRelativePathPolicy.normalize(path)
            if (!normalized.isSuccess()) return null
            when {
                types[index] == 2 -> {
                    if (sizes[index] != 0L || ordinals[index] != -1 || !zeroDigest(digests, index)) {
                        return null
                    }
                    val entry = Entry(
                        normalized.normalized!!,
                        NarGhostTreePolicy.Type.DIRECTORY,
                        0,
                        -1,
                        null,
                    )
                    inventory[entry.path()!!] = entry
                    policy.add(NarGhostTreePolicy.InputEntry.directory(path))
                }

                types[index] == 1 && sizes[index] >= 0 && ordinals[index] >= 0 &&
                    ordinals[index] < count && !seen[ordinals[index]] -> {
                    seen[ordinals[index]] = true
                    files++
                    val digest = digests.copyOfRange(
                        index * DIGEST_BYTES,
                        (index + 1) * DIGEST_BYTES,
                    )
                    val entry = Entry(
                        normalized.normalized!!,
                        NarGhostTreePolicy.Type.FILE,
                        sizes[index],
                        ordinals[index],
                        digest,
                    )
                    inventory[entry.path()!!] = entry
                    policy.add(NarGhostTreePolicy.InputEntry.file(path, sizes[index], digest))
                }

                else -> return null
            }
        }
        for (ordinal in 0 until files) {
            if (!seen[ordinal]) return null
        }
        for (ordinal in files until seen.size) {
            if (seen[ordinal]) return null
        }
        return Prepared(policy, inventory)
    }

    private fun zeroDigest(values: ByteArray, index: Int): Boolean {
        val start = index * DIGEST_BYTES
        for (offset in 0 until DIGEST_BYTES) {
            if (values[start + offset].toInt() != 0) return false
        }
        return true
    }

    private fun identity(device: Long, inode: Long): ByteArray = ByteArray(16).also {
        putLong(it, 0, device)
        putLong(it, 8, inode)
    }

    private fun putLong(target: ByteArray, offset: Int, initialValue: Long) {
        var value = initialValue
        for (index in 7 downTo 0) {
            target[offset + index] = value.toByte()
            value = value ushr 8
        }
    }
}
